flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/EdgeManagerBuildUtil.java

```diff
@@ -197,10 +202,22 @@ private static ConsumedPartitionGroup createAndRegisterConsumedPartitionGroupToE
         ConsumedPartitionGroup consumedPartitionGroup =
                 ConsumedPartitionGroup.fromMultiplePartitions(
                         numConsumers, consumedPartitions, intermediateResult.getResultType());
+        finishConsumablePartitions(intermediateResult, consumedPartitions, consumedPartitionGroup);
         registerConsumedPartitionGroupToEdgeManager(consumedPartitionGroup, intermediateResult);
         return consumedPartitionGroup;
     }
 
+    private static void finishConsumablePartitions(
+            IntermediateResult intermediateResult,
+            List<IntermediateResultPartitionID> consumedPartitionId,
+            ConsumedPartitionGroup consumedPartitionGroup) {
+        for (IntermediateResultPartitionID intermediateResultPartitionID : consumedPartitionId) {
+            if (intermediateResult.getPartitionById(intermediateResultPartitionID).isConsumable()) {
```

Comment:
And it would be better if you add a comment to tell this is for dynamic graph

Reply:
We will make hybrid edge also can be mark finished to support speculative, so i think it can be changed to `!mustBePipelinedConsumed && isConsumable`.

BTW, `hasDataProduced` will not be set to true in the pipelined edge now (because the`markDataProduced()` method has no production code caller). Can we directly change this status to `hasDataAllProduced`, and rename `isConsumable()` to `isAllDataProduced ()`. WDYT?

Reply:
`!mustBePipelinedConsumed && isConsumable` is ok to me.

I think it's a good idea to change to `hasDataAllProduced`, becase the `consumable` is ambiguous. But let's wait for @zhuzhurk review as well (since the `consumable` was originally introduced by his change).